File: server/heartbeat/sources/registry.js
// The entire plug-in surface for the Heartbeat — pure, zero imports, same
// discipline as personality.js's detectFloors()/self-signals.js. Adding a
// future source (calendar, business monitoring, ...) is one call to
// registerSource() from wherever that source's own module lives; nothing
// here needs to change, and nothing here knows what any source actually
// checks.
//
// A source is `{ id, defaultIntervalMs, listItems(), check(itemKey) }`:
//   - `id` — a short, stable string (e.g. 'jobs', 'commitments'), used as
//     heartbeat_schedule.source_id.
//   - `defaultIntervalMs` — how often a newly-seen item of this source's own
//     should be checked, unless the source itself gives a per-item override.
//   - `listItems()` — async, returns `[{itemKey, intervalMs?}]`, the CURRENT
//     set of things this source wants watched (dynamic — e.g. every job
//     currently holding an undelivered Tier 1 outbox row). engine.js
//     reconciles this against the persisted schedule every tick: new keys
//     get a row, keys no longer listed get pruned.
//   - `check(itemKey)` — async, returns `{finding, checkState?}`. `finding`
//     is `{summary, detail?}` or `null` (nothing to report right now).
//     `checkState` is optional free-form JSON a source wants remembered
//     between ticks for this exact item (e.g. commitments-source.js's own
//     cached parsed deadline) — omit it entirely to leave the persisted
//     value untouched; pass `null` to explicitly clear it. Read back via
//     schedule-store.js's getItem(sourceId, itemKey). Thrown errors are
//     caught by engine.js, never by the source itself — see engine.js's own
//     per-check isolation.

const sources = new Map();

export function registerSource(source) {
  if (!source?.id) throw new Error('A Heartbeat source needs an id.');
  if (typeof source.listItems !== 'function' || typeof source.check !== 'function') {
    throw new Error(`Heartbeat source "${source.id}" must implement listItems() and check()`);
  }
  sources.set(source.id, source);
}

export function listSources() {
  return [...sources.values()];
}

/** Test-only escape hatch — clears every registered source so a scratch test process can register its own fakes with no leftover state. Production code never calls this. */
export function _resetForTests() {
  sources.clear();
}

File: backend/jarvis/heartbeat/sources/registry.py
"""The whole plug-in surface for something worth watching.

Pure and zero-import on purpose: the engine that ticks and the layer that judges
urgency know nothing about what any source actually watches, and a new one is a
single `register()` call rather than a change to either.

A source is:

* `id` — stable; it becomes the schedule row's key and the trace's ref.
* `default_interval_ms` — how often its items are worth checking.
* `list_items()` — what it is currently watching, as `{"itemKey", "intervalMs"?}`.
* `check(item_key)` — `{"finding": {...} | None, "checkState": ... }`.

**A source whose condition can stay true across many ticks must keep its own
"already reported" memory in `checkState`.** A tier-3 verdict creates no outbox
row, so the broker's own duplicate check has nothing to see, and the same
unresolved situation will otherwise produce a fresh finding, a fresh model call
and a fresh notification on every single tick, forever. That is not a
hypothetical: it is the bug this contract exists to prevent.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable


@dataclass(frozen=True)
class Source:
    id: str
    default_interval_ms: int
    list_items: Callable[[], list[dict[str, Any]]]
    check: Callable[[str], dict[str, Any]]


_sources: dict[str, Source] = {}


def register(source: Source) -> None:
    if not getattr(source, "id", None):
        raise ValueError("a heartbeat source needs an id")
    if not callable(source.list_items) or not callable(source.check):
        raise TypeError(f'source "{source.id}" must implement list_items() and check()')
    _sources[source.id] = source


def list_sources() -> list[Source]:
    return list(_sources.values())


def get_source(source_id: str) -> Source | None:
    return _sources.get(source_id)


def reset() -> None:
    _sources.clear()
